File: src/fastica.rs
use super::contrast::*;
use super::data::*;
use arrayfire::*;

/**
 * *** FastICA ***
 *
 *  h_white_matrix:
 *      whitened matrix reference - preprocessing technique is chosen by the user
 *      this crate provides whitening via pca or zca.
 *  
 *  max_iter:
 *      maximum number of iterations
 *
 *  n_components:
 *      number of components desired
 *
 *  conv_threshold:
 *      maximum threshold to accept convergence
 *
 * alpha:
 *      argument for g and g' functions
 *
 * ini_weights:
 *      initial weight matrix
 *   
 */
pub fn fast_ica(
    matrix: &Matrix,
    n_components: u64,
    max_iter: u64,
    conv_threshold: f64,
    alpha: f64,
    cfid: ContrastFunctionId,
) -> Matrix {

    let cfunc = get_contrast_function(cfid);

    let mut ret_weights = empty_matrix(n_components, n_components);

    for comp_i in 0..n_components {
        let mut wp = col(&matrix, comp_i);
        
        // decorrelation
        if comp_i > 1 {
            wp = decorrelate(&wp, &ret_weights, comp_i);
        }
        wp = normalize(&wp);
        
        let converged = false;

        while !converged {
            let wx = matmul(&wp, &matrix, MatProp::TRANS, MatProp::NONE);

        }      
    }

    unimplemented!()
}

    
/*
    
    t = zeros(size(w1))
    for u = 1:(i-1)
        k = sum(w1 .* retW[u,:,])
        t = t + k * retW[u,:,]
    w1 - t
*/

fn decorrelate(col_mat: &Matrix, ret_w: &Matrix, iter: u64) -> Matrix {
    let mut t = Matrix::new_empty(col_mat.dims());
    
    for col_i in 0..iter {
        let rw_col_i = col(ret_w, col_i);
        
        let k = { 
            let temp = mul(col_mat, &rw_col_i, true); 
            sum(&temp,0) 
        };
        
        t = {
            let temp = matmul(&k, col_mat, MatProp::NONE, MatProp::NONE);
            add(&t, &temp, true)
        };

    }

    sub(col_mat, &t, false)
}

fn normalize(matrix: &Matrix) -> Matrix {
    
    let matrix_dims = matrix.dims();
    
    let ini: Matrix = Matrix::new_empty(matrix_dims);
    
    let reducer = |mat, idx| { 
            let col_i     = col(matrix, idx);
            let col_norm  = norm(&col_i, NormType::VECTOR_1, 0.0, 0.0);
            let new_col   = div(&col_i, &col_norm, true);
            set_col(&mat, &new_col, idx) 
    };

    (0..matrix_dims.get()[0]).fold(ini, reducer)
}

fn distance(w: &Matrix, nw: &Matrix) -> f64 {
    /*
       dist = w1 .* wp;
       dist = sum(dist)
       dist = abs. (dist) .- 1.0
       dist = abs. (dist)
       return maximum(dist)
    */

    let dist = {
        let w_prod = mul(w, nw, true);

        let vec_size = w.dims()[0] as usize;

        let mut buffer = empty_row_vector(w.dims()[0]);

        for i in 0..vec_size {
            buffer = {
                let col_sum = sum(&w_prod, i as i32);
                set_col(&buffer, &col_sum, i as u64)
            };
        }

        let temp = abs(&buffer);
        let temp = sub(&temp, &1.0, true);
        abs(&temp)
    };

    let (max_val, _, _) = imax_all(&dist);
    max_val 
}

fn update_weights(
    weights: &Matrix,
    whitened_mat: &Matrix,
    cf: &ContrastFunc,
    alpha: f64,
) -> Matrix {
    let (g, dg) = {
        let temp = dot(weights, whitened_mat, MatProp::TRANS, MatProp::NONE);
        cf(&temp, alpha)
    };
    let new_weights = {
        let temp = mul(whitened_mat, &g, false);
        temp
        // TODO finish weight updating
    };
    new_weights
}

File: src/fastica.cpp
#include "../include/fastica.h"
#include <iostream>
#include <time.h>

using namespace Eigen;
using namespace contrast;
using namespace std;

double conv_distance(const RowVectorXd& w, const RowVectorXd& nw)
{
    return abs(abs(w.cwiseProduct(nw).sum()) - 1.0);
}

VectorXd decorrelate(const VectorXd &row_mat, const MatrixXd &ret_w, size_t n_comps)
{

    VectorXd t = VectorXd::Constant(ret_w.cols(), 0.0);

    for (size_t i = 0; i < n_comps - 1; ++i)
    {

        VectorXd rw_row_i = ret_w.row(i).transpose();

        double k = row_mat.cwiseProduct(rw_row_i).sum();

        t += k * rw_row_i;
    }

    return row_mat - t;
}

MatrixXd *fast_ica_impl(
    const MatrixXd &matrix,
    const MatrixXd &ini_weights,
    unsigned int n_components,
    double conv_threshold,
    double alpha,
    ContrastFunction contrast_function,
    bool verbose,
    unsigned int max_iter)
{

    auto ret_weights = new MatrixXd(n_components, n_components);
    *ret_weights = MatrixXd::Zero(n_components, n_components);

    for (size_t comp_i = 0; comp_i < n_components; ++comp_i)
    {

        VectorXd wp = ini_weights.row(comp_i).transpose();

        if (verbose)
            cout << "Begin search for component " << comp_i << endl;

        if (comp_i > 0) {
            wp = decorrelate(wp, *ret_weights, comp_i);
        }
        wp = wp.normalized();


        for (size_t iter_i = 0;; ++iter_i) {

            if (verbose)
                cout << "Begin iteration " << iter_i << endl;

            MatrixXd wx = wp.transpose() * matrix;
            wx.eval();

            MatPair dxg_pair = contrast_function(wx, alpha);
            MatrixXd dg = get<0>(dxg_pair).array().colwise().replicate(2);
            MatrixXd d2g = get<1>(dxg_pair);

            MatrixXd xdg = matrix.cwiseProduct(dg);

            MatrixXd v1 = xdg.rowwise().mean();

            VectorXd v2 = d2g.mean() * wp;

            VectorXd nw = v1 - v2;


            if (comp_i > 0) {
                nw = decorrelate(nw, *ret_weights, comp_i);
            }
            nw = nw.normalized();


            double dist = conv_distance(wp, nw);

            if (verbose)
                cout << "distance: " << dist << endl;

            wp = nw;

            if (dist < conv_threshold || iter_i >= max_iter)
            {
                cout << "converged in iteration: " << iter_i << endl;

                ret_weights->row(comp_i) = wp;
                break;
            }
        }
    }

    return ret_weights;
}

// TODO implement
MatrixXd *fastica::fast_ica(
    const Eigen::MatrixXd &dataset,
    const Eigen::MatrixXd *white_matrix,
    const Eigen::MatrixXd *weights,
    ICA_Params parameters)
{

    if (parameters.alpha < 1.0 && parameters.alpha > 2.0) {
        cerr << "alpha must be between 1.0 and 2.0" << endl;
        return nullptr;
    }

    MatrixXd ws;
    if(weights == nullptr) {
        srand(parameters.seed != nullptr ? *parameters.seed : ((unsigned int)time(nullptr)));
        ws = MatrixXd::Random(parameters.n_components, parameters.n_components);
    } else {
        ws = *weights;
    }

    MatrixXd wt = white_matrix != nullptr ? white_matrix->transpose() : dataset;
    cout <<"wt\n\n" << wt << endl;
    MatrixXd dt = dataset * wt; // TODO whitening
    cout <<"dt\n\n" << dt << endl;
    return fast_ica_impl(
        dt.transpose(),
        ws,
        parameters.n_components,
        parameters.conv_threshold,
        parameters.alpha,
        get_contrast_function(parameters.cont_func_id),
        parameters.verbose,
        parameters.max_iter);
}
